File: Abstract Factory/Exemplos/Ex2_Combos/Ex2.js
/**
 * PADRÃO: ABSTRACT FACTORY (FÁBRICA ABSTRATA)
 * 
 * Este exemplo mostra como criar combos de produtos relacionados (Moda e Eletrônico)
 * garantindo consistência entre eles através de fábricas específicas.
 */

// 1. PRODUTOS ABSTRATOS
class ProdutoModa {
    exibirDetalhes() { throw new Error("Método abstrato"); }
}

class ProdutoEletronico {
    exibirDetalhes() { throw new Error("Método abstrato"); }
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA FASHION
class CamisaFashion extends ProdutoModa {
    exibirDetalhes() { return "Camisa Fashion: Estampa moderna."; }
}

class CelularFashion extends ProdutoEletronico {
    exibirDetalhes() { return "Celular Fashion: Design elegante."; }
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA TECH
class CamisaTech extends ProdutoModa {
    exibirDetalhes() { return "Camisa Tech: Tecido inteligente."; }
}

class CelularTech extends ProdutoEletronico {
    exibirDetalhes() { return "Celular Tech: Alta performance."; }
}

// 3. FABRICA ABSTRATA
class FabricaCombo {
    criarCamisa() { throw new Error("Método abstrato"); }
    criarCelular() { throw new Error("Método abstrato"); }
}

// 4. FABRICAS CONCRETAS
class FabricaComboFashion extends FabricaCombo {
    criarCamisa() { return new CamisaFashion(); }
    criarCelular() { return new CelularFashion(); }
}

class FabricaComboTech extends FabricaCombo {
    criarCamisa() { return new CamisaTech(); }
    criarCelular() { return new CelularTech(); }
}

// 5. CLIENTE
function lojaVirtual(fabrica) {
    const camisa = fabrica.criarCamisa();
    const celular = fabrica.criarCelular();

    console.log(camisa.exibirDetalhes());
    console.log(celular.exibirDetalhes());
}

// Testando os combos
console.log("Combo Fashion:");
lojaVirtual(new FabricaComboFashion());

console.log("-------");

console.log("Combo Tech:");
lojaVirtual(new FabricaComboTech());

File: Abstract Factory/Exemplos/Ex2_Combos/Ex2.java
/**
 * PADRÃO: ABSTRACT FACTORY (FÁBRICA ABSTRATA)
 * 
 * Exemplo de uma loja que vende combos de produtos (Moda e Eletrônico).
 * Cada fábrica (Fashion ou Tech) garante que os produtos do combo combinem entre si.
 */

// 1. PRODUTOS ABSTRATOS
interface ProdutoModa {
    String exibirDetalhes();
}

interface ProdutoEletronico {
    String exibirDetalhes();
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA FASHION
class CamisaFashion implements ProdutoModa {
    public String exibirDetalhes() { return "Camisa Fashion: Estampa moderna e tecido leve."; }
}

class CelularFashion implements ProdutoEletronico {
    public String exibirDetalhes() { return "Celular Fashion: Design elegante com acabamento em vidro."; }
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA TECH
class CamisaTech implements ProdutoModa {
    public String exibirDetalhes() { return "Camisa Tech: Tecido inteligente com controle térmico."; }
}

class CelularTech implements ProdutoEletronico {
    public String exibirDetalhes() { return "Celular Tech: Processador de última geração e tela 120Hz."; }
}

// 3. FABRICA ABSTRATA
interface FabricaCombo {
    ProdutoModa criarCamisa();
    ProdutoEletronico criarCelular();
}

// 4. FABRICAS CONCRETAS
class FabricaComboFashion implements FabricaCombo {
    public ProdutoModa criarCamisa() { return new CamisaFashion(); }
    public ProdutoEletronico criarCelular() { return new CelularFashion(); }
}

class FabricaComboTech implements FabricaCombo {
    public ProdutoModa criarCamisa() { return new CamisaTech(); }
    public ProdutoEletronico criarCelular() { return new CelularTech(); }
}

// 5. CLIENTE
public class Ex2 {
    public static void lojaVirtual(FabricaCombo fabrica) {
        ProdutoModa camisa = fabrica.criarCamisa();
        ProdutoEletronico celular = fabrica.criarCelular();

        System.out.println(camisa.exibirDetalhes());
        System.out.println(celular.exibirDetalhes());
    }

    public static void main(String[] args) {
        System.out.println("Combo Fashion:");
        lojaVirtual(new FabricaComboFashion());
        
        System.out.println("-------");
        
        System.out.println("Combo Tech:");
        lojaVirtual(new FabricaComboTech());
    }
}
